### ISSUE TITLE
Hostnames not resolving breaks some features of the cluster

### ISSUE THREAD

**Issue by axinojolais** [juju-solutions/bundle-canonical-kubernetes#203]
Hi,

In a typical OpenStack cluster, and as far as I know, the hostname of the VMs will not resolve in DNS. This is causing issues for certain commands (notably "kubectl exec" and "kubectl logs") which expect the hostnames of the kuebernetes-worker units to be resolve on the kubernetes-master units.

The workaround we're using is to add the IP and hostnames to /etc/hosts on the kubernetes-master units. This workaround, even though dirty, could make its way into the kubernetes-master charm : since it's already related to the kubernetes-worker application, it shouldn't be too hard to gather all the IPs / hostnames and stash them in /etc/hosts.

I don't have another solution to propose, but perhaps you do ?

Thanks

**Comment by chuckbutler** [comment]
Hostname hacks are messy, as they can wind up leaving stale entries behind, get out of sync after reboots and other variations of this. We piloted a similar approach in the big data charms a long time ago and it did create a lot of problems. 

I'm not entirely opposed to finding a working solution for the OpenStack users out there, but ideally, FQDN should be a thing every cloud supports. I realize that's a big ask. There was also talk of having juju support this natively, however, I haven't looked at that in quite some time and I'm not even positive it landed outside of the LXD scope.

I think we need to revisit this line item and hash this out for a roadmap discussion or find a suitable way to provide cluster-level hostname DNS out of band of the cloud provider (perhaps with our own dnsmasq workaround provided by a charm running on the masters providing another set of DNS, but we have to be cognizant of the fact there's a limit to how many dns servers a host can have and I imagine that many cloud providers are already filling this list up before we can go stuffing our noses in the resolver)

**Comment by arosales** [comment]
@axinojolais 

To confirm your environment are you running in any containers and what is the Juju version you are using.

-thanks,
Antonio

**Comment by axinojolais** [comment]
@arosales this is running on KVM instances, no container involved. And this is deployed with juju 2 (currently 2.0.2)

**Comment by chuckbutler** [comment]
related to https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/171

closing in favor of the original bug which documents this pretty extensively.
